File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.cpp
class Solution 
{
    public:
    int minSetSize(vector<int>& arr) 
    {
        unordered_map<int, int> m;
        vector<int> vec;
        int s=0, l=arr.size();
        
        for(int i:arr)
            m[i]++;
        arr.clear();
        for(auto it:m)
            arr.push_back(it.second);
        
        sort(arr.begin(), arr.end(), greater<int>());
               
        for(int i=0; i<arr.size(); i++)
        {
            if((s+arr[i]) >= (l/2))
                return i+1;
            else
                s += arr[i];
        }
        return 0;
    }
};

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution 
{
    public int minSetSize(int[] arr) 
    {
        Arrays.sort(arr);
        ArrayList<Integer> freq = new ArrayList<Integer>();
        int l = arr.length, count=1, s=0;
        
        for(int i=0; i<l-1; i++)
        {
            if(arr[i] == arr[i+1])
                count++;
            else
            {
                freq.add(count);
                count = 1;
            }
        }
        if(arr[l-1] == arr[l-2])
            freq.add(count);
        else if(arr[l-1] != arr[l-2])
            freq.add(1);            
        
        Collections.sort(freq, Collections.reverseOrder());
        for(int i=0; i<freq.size(); i++)
        {
            s += freq.get(i);
            if(s >= (l/2))
                return i+1;
        }
        return 0;          
    }
}